pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go

```diff
@@ -0,0 +1,280 @@
+func (c *cell) GetCellBoundChunk() []uint32 {
+	if c.threadsPerCore <= len(c.smallChunks) {
+		selected := c.smallChunks[:c.threadsPerCore]
+		c.smallChunks = c.smallChunks[c.threadsPerCore:]
+		return selected
+	}
+	return nil
+}
+
+// GetCellBoundChunk will allocate as many threads as possible
+// and return them, even if it can only satisfy parts of the request.
+func (c *cell) GetCellBoundChunkUpTo(threads int) []uint32 {
+	selector := threads
+	if threads > len(c.smallChunks) {
+		selector = len(c.smallChunks)
+	}
+	selected := c.smallChunks[:selector]
+	c.smallChunks = c.smallChunks[selector:]
+	return selected
+}
+
+// GetThread will first try to allocate a thread from fragmented cores
+// but fall back to not fragmented cores if the request can't be satisfied otherwise
+func (c *cell) GetThread() *uint32 {
+	if len(c.smallChunks) > 0 {
+		thread := c.smallChunks[0]
+		c.smallChunks = c.smallChunks[1:]
+		return &thread
+	} else if len(c.bigThreadChunks) > 0 {
+		thread := c.bigThreadChunks[0][0]
+		remaining := c.bigThreadChunks[0][1:]
+		if len(remaining) >= c.threadsPerCore {
+			c.bigThreadChunks[0] = remaining
+		} else {
+			c.bigThreadChunks = c.bigThreadChunks[1:]
+			c.smallChunks = append(c.smallChunks, remaining...)
+		}
+		return &thread
+	}
+	return nil
+}
+
+func (c *cell) IsEmpty() bool {
+	return len(c.smallChunks) == 0 && len(c.bigThreadChunks) == 0
+}
+
+type cpuPool struct {
+	// cells contains a host thread mapping of host threads to their cores
+	// and cores to their numa cells
+	cells []*cell
+	// threadsPerCore is the amount of vcpu threads per vcpu core
+	threadsPerCore int
+	// cores is the amount of vcpu cores requested by the VMI
+	cores int
+	// allowCellCrossing allows inefficient cpu mapping where a single
+	// core can have threads on different host numa cells
+	allowCellCrossing bool
+	// availableThreads is the amount of all threads assigned to the pod
+	availableThreads int
+}
+
+func NewStrictCPUPool(requestedToplogy *api.CPUTopology, nodeTopology *v1.Topology, cpuSet []int) VCPUPool {
+	return newCPUPool(requestedToplogy, nodeTopology, cpuSet, false)
+}
+
+func NewRelaxedCPUPool(requestedToplogy *api.CPUTopology, nodeTopology *v1.Topology, cpuSet []int) VCPUPool {
+	return newCPUPool(requestedToplogy, nodeTopology, cpuSet, true)
+}
+
+func newCPUPool(requestedToplogy *api.CPUTopology, nodeTopology *v1.Topology, cpuSet []int, allowCellCrossing bool) *cpuPool {
+	pool := &cpuPool{threadsPerCore: int(requestedToplogy.Threads), cores: int(requestedToplogy.Cores * requestedToplogy.Sockets), allowCellCrossing: allowCellCrossing, availableThreads: len(cpuSet)}
+	cores := cpuChunksToCells(cpuSet, nodeTopology)
+
+	for _, coresOnCell := range cores {
+		c := cell{threadsPerCore: int(requestedToplogy.Threads)}
+		for j, core := range coresOnCell {
+			if len(core) >= c.threadsPerCore {
+				c.bigThreadChunks = append(c.bigThreadChunks, coresOnCell[j])
```

Comment:
perhaps we didn't need to make this generic and simply call this fullCoreThreads and s/smallChunks/partialCoreThreads all over the code - this would make it more clear from my pov.
There are no real systems that have more than 2 hyper-threads and if it's a virtual system is pretty much not useful if the guest threads are not pinned directly to the L0 host - but there are no L0 hosts with threads > 2... maybe we should just reject cases where threads > 2